crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,811 @@
+        // NOTE: Block doesn't have a parent, and if we receive this block in `make_canonical`
+        // function this could be a trigger to initiate p2p syncing, as we are missing the
+        // parent.
+        Ok(false)
+    }
+
+    /// Do finalization of blocks. Remove them from tree
+    pub fn finalize_block(&mut self, finalized_block: BlockNumber) {
+        let mut remove_chains = self.block_indices.finalize_canonical_blocks(finalized_block);
+
+        while let Some(chain_id) = remove_chains.pop_first() {
+            if let Some(chain) = self.chains.remove(&chain_id) {
+                remove_chains.extend(self.block_indices.remove_chain(&chain));
+            }
+        }
+    }
+
+    /// Update canonical hashes. Reads last N canonical blocks from database and update all indices.
+    pub fn update_canonical_hashes(
+        &mut self,
+        last_finalized_block: BlockNumber,
+    ) -> Result<(), Error> {
+        self.finalize_block(last_finalized_block);
+
+        let num_of_canonical_hashes =
+            self.finalization_window + self.block_indices.num_of_additional_canonical_block_hashes;
+
+        let last_canonical_hashes = self
+            .externals
+            .db
+            .tx()?
+            .cursor_read::<tables::CanonicalHeaders>()?
+            .walk_back(None)?
+            .take(num_of_canonical_hashes as usize)
+            .collect::<Result<BTreeMap<BlockNumber, BlockHash>, _>>()?;
+
+        let mut remove_chains = self.block_indices.update_block_hashes(last_canonical_hashes);
+
+        // remove all chains that got discarded
+        while let Some(chain_id) = remove_chains.first() {
+            if let Some(chain) = self.chains.remove(chain_id) {
+                remove_chains.extend(self.block_indices.remove_chain(&chain));
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Make block and its parent canonical. Unwind chains to database if necessary.
+    ///
+    /// If block is alreadt
+    pub fn make_canonical(&mut self, block_hash: &BlockHash) -> Result<(), Error> {
+        let chain_id = if let Some(chain_id) = self.block_indices.get_blocks_chain_id(block_hash) {
+            chain_id
+        } else {
+            if self.block_indices.is_block_hash_canonical(block_hash) {
+                // If block is already canonical don't return error.
+                return Ok(())
+            }
+            return Err(ExecError::BlockHashNotFoundInChain { block_hash: *block_hash }.into())
+        };
+        let chain = self.chains.remove(&chain_id).expect("To be present");
+
+        // we are spliting chain as there is possibility that only part of chain get canonicalized.
+        let (canonical, pending) = chain.split_at_block_hash(block_hash);
+        let canonical = canonical.expect("Canonical chain is present");
+
+        if let Some(pending) = pending {
+            // fork is now canonical and latest.
+            self.block_indices.insert_chain(chain_id, &pending);
```

Comment:
is this going to override the existing values for `chain_id`? or do we need to delete them ourselves?

Reply:
Yeah, chain_id can be reused as practicaly it is same chain, and just part of it got canonicalized.
Reworded the comment:
` // rest of splited chain is inserted back with same chain_id.`